### DISCUSSION TITLE
Lazy loading an addin

### DISCUSSION THREAD

**Discussion by savornicesei** [cake-build/cake#3420] [Q&A]
Hi all,

I would like to load and use a cake addin that's build in/by the same cake script.
I suppose I could do something similar to [this](https://stackoverflow.com/questions/46648455/conditionally-load-cake-files): have the 'main' script generate a tmp.cake with the path to addin (addin path is known only at build time) and a second script that calls the tmp one and does it's magic.

Could there be a nicer/simpler way of doing it?

Thanks,
Simo

**Comment by pascalberger** [comment]
What's your use case? Running integration tests in a build pipeline for an addin?

**Reply by pascalberger** [reply]
@savornicesei The integration test feature of [Cake.Recipe](https://github.com/cake-contrib/Cake.Recipe) might be an interesting example for you.

[Here](https://github.com/cake-contrib/Cake.Issues.GitRepository/blob/develop/tests/integration/tests.cake) is an example repo which uses it to run a Cake Script, after it has built an addin to test the addin in a script: 

```
#addin nuget:?package=Cake.Issues&prerelease
#addin nuget:?package=Cake.Issues.Reporting&prerelease
#addin nuget:?package=Cake.Issues.Reporting.Generic&prerelease
#reference "../../tools/Addins/Cake.Issues.GitRepository/lib/netstandard2.0/Cake.Issues.GitRepository.dll"
```

[Here](https://github.com/cake-contrib/Cake.Recipe/blob/3842b2b7f5d2176130e2814eccc511f14e1e2edf/Source/Cake.Recipe/Content/testing.cake#L207) is the corresponding task in Cake.Recipe which runs the above script:

```
BuildParameters.Tasks.IntegrationTestTask = Task("Run-Integration-Tests")
    .WithCriteria(() => BuildParameters.ShouldRunIntegrationTests, "Cake script integration tests have been disabled")
    .IsDependentOn("Default")
    .Does(() =>
    {
            CakeExecuteScript(BuildParameters.IntegrationTestScriptPath,
                new CakeSettings
                {
                    Arguments = new Dictionary<string, string>
                    {
                        { "verbosity", Context.Log.Verbosity.ToString("F") }
                    }
                });
    });
```

**Comment by nils-a** [comment]
In the end, you'd only need to ensure that the tool is available somewhere in the path or in the tools folder.
You could use [`CakeExecuteExpression` or `CakeExecuteScript`](https://cakebuild.net/api/Cake.Common.Tools.Cake/CakeAliases/) to let Cake do it's thing.
Alternatively you could probably utilize [`NuGetInstall`](https://cakebuild.net/api/Cake.Common.Tools.NuGet/NuGetAliases/) to install the tool manually in the tools folder.

**Reply by savornicesei** [reply]
@pascalberger  
I've been working on [Wyam2](https://github.com/Wyam2/wyam) - fork of Wyam. Obvious its documentation is generated by itself but it requires some manual work after the released NuGet packages are available on NuGet.
What I want for the [docs](https://github.com/Wyam2/docs) repo is to:
- get Wyam2 code from GitHub or from a local folder
- build it
- use Cake.Wyam to generate docs
-  push them to github.io repo for publish

@nils-a 
The problem is that the cake script is evaluated first - when the adddin is not loaded - so of course it complains when calling it. I guess an additional cake script might do the trick with `CakeExecuteScript` but I would love to keep everything in a single file - if it's possible.

**Comment by devlead** [comment]
Don't know if it helps, but you can control the location of tools, addins, and modules by specifying environment variables.
Example from our integration tests.

https://github.com/cake-build/cake/blob/80c431f07577faa0482a131509f6f23e4d53a483/tests/integration/Cake.Core/Scripting/AddinDirective.cake#L18-L27
